python/ray/ml/train/integrations/horovod/horovod_trainer.py

```diff
@@ -0,0 +1,196 @@
+    the "train" key), then it will be split into multiple dataset
+    shards that can then be accessed by ``ray.train.get_dataset_shard("train")`` inside
+    ``train_loop_per_worker``. All the other datasets will not be split and
+    ``ray.train.get_dataset_shard(...)`` will return the the entire Dataset.
+
+    Inside the ``train_loop_per_worker`` function, you can use any of the
+    :ref:`Ray Train function utils <train-api-func-utils>`.
+
+    .. code-block:: python
+
+        def train_loop_per_worker():
+            # Report intermediate results for callbacks or logging.
+            train.report(...)
+
+            # Checkpoints the provided args as restorable state.
+            train.save_checkpoint(...)
+
+            # Returns dict of last saved checkpoint.
+            train.load_checkpoint()
+
+            # Returns the Ray Dataset shard for the given key.
+            train.get_dataset_shard("my_dataset")
+
+            # Returns the total number of workers executing training.
+            train.get_world_size()
+
+            # Returns the rank of this worker.
+            train.get_world_rank()
+
+            # Returns the rank of the worker on the current node.
+            train.get_local_rank()
+
+    You could use ``TensorflowPredictor`` or ``TorchPredictor`` in conjunction with
+    HorovodTrainer. You must save it under the "model" kwarg in
+    ``train.save_checkpoint()``.
+
+    Example:
+
+    .. code-block:: python
+
+        import horovod.torch as hvd
+        import torch
+        import torch.nn as nn
+
+        import ray
+        from ray import train
+        from ray.ml.train.integrations.torch import TorchTrainer
+
+
+        input_size = 1
+        layer_size = 15
+        output_size = 1
+        num_epochs = 3
+
+        class NeuralNetwork(nn.Module):
+            def __init__(self):
+                super(NeuralNetwork, self).__init__()
+                self.layer1 = nn.Linear(input_size, layer_size)
+                self.relu = nn.ReLU()
+                self.layer2 = nn.Linear(layer_size, output_size)
+
+            def forward(self, input):
+                return self.layer2(self.relu(self.layer1(input)))
+
+
+        def train_loop_per_worker():
+            use_adasum = False
+            hvd.init()
+            dataset_shard = train.get_dataset_shard("train")
+            model = NeuralNetwork()
+            device = train.torch.get_device()
+            model.to(device)
+            loss_fn = nn.MSELoss()
+
+            # By default, Adasum doesn't need scaling up learning rate.
+            lr_scaler = hvd.size() if not use_adasum else 1
+            optimizer = torch.optim.SGD(model.parameters(), lr=0.1 * lr_scaler)
+
+            # Horovod: wrap optimizer with DistributedOptimizer.
+            optimizer = hvd.DistributedOptimizer(
+                optimizer,
+                named_parameters=model.named_parameters(),
+                op=hvd.Adasum if use_adasum else hvd.Average,
```

Comment:
if `use_adasum` is always False maybe we can just remove all the adasum specific logic?